### ISSUE TITLE
T_OPEN_TAG with white character

### ISSUE THREAD

**Issue by webimpress** [squizlabs/PHP_CodeSniffer#1978]
Content of T_OPEN_TAG content is always with space or new line character (or maybe something else).
Why we don't have there two separate tokens T_OPEN_TAG and T_WHITESPACE?

I know that there MUST be a white character after `<?php` for a valid php file, but it could be either space or new line (or tab...?).
I was trying to solve it by adding:
```php
            /*
                Check if T_OPEN_TAG contains also white character
            */

            if ($tokenIsArray === true
                && $token[0] === T_OPEN_TAG
                && preg_match('/\s+$/i', $token[1], $matches)
            ) {
                $token[1] = rtrim($token[1]);

                $finalTokens[$newStackPtr] = self::standardiseToken($token);
                $newStackPtr++;

                $newToken            = [];
                $newToken['code']    = T_WHITESPACE;
                $newToken['type']    = 'T_WHITESPACE';
                $newToken['content'] = $matches[0];
                $finalTokens[$newStackPtr] = $newToken;

                $newStackPtr++;
                continue;
            }
```

in php tokenizer `PHP.php` but after then many tests failed.

Is it done like that on purpose or should we change it and have there two tags instead of one?
I think it would be nice to have two separate, but maybe it will be BC Break? What do you think about changing it in v4?

**Comment by gsherwood** [comment]
> Why we don't have there two separate tokens T_OPEN_TAG and T_WHITESPACE?

Because that's how PHP tokenizes those tags. The same is true with close tags. For example, this content:
```php
<?php
echo "foo";
?>
```
Is tokenized by PHP like this:
```
0 : T_OPEN_TAG => <?php\n
1 : T_ECHO => echo
2 : T_WHITESPACE =>
3 : T_CONSTANT_ENCAPSED_STRING => "foo"
4 : ;
5 : T_WHITESPACE => \n
6 : T_CLOSE_TAG => ?>\n
```
I have no idea what the reason is, but it has always been that way and has never caused me any problems, so I don't want to change that unless there is a really good reason.

Do you have a good reason that this needs to change or is it just a suggestion given it looks strange?

**Comment by webimpress** [comment]
@gsherwood It was just suggestion to split it to have clear tokens.
It's a bit confusing sometimes, you have to remember about it on writing sniffs, because if we have:
```php
<?php   echo 'three spaces before';
```
it is tokenized as:
```
0 : T_OPEN_TAG => <?php_ // with one space
1 : T_WHITESPACE => __ // two spaces
...
```

I think just for consistency and clarity it could be changed in v4, as now many sniffs can rely on that 'extra' white character there. What do you think?

**Comment by jrfnl** [comment]
For what it's worth: I don't think this should be changed. A PHP open tag is not valid unless it has whitespace after it, be it a new line or a space. That's part of the language definition and not something which should be redefined by PHPCS or even defined differently by PHPCS.
